### PR TITLE
ENH: linalg: Add wrapper for `?tgsyl`

### LINKED ISSUE #18404: ENH: Add wrapper for LAPACK functions stgsyl and dtgsyl
### Is your feature request related to a problem? Please describe.

I need the LAPACK functions ```stgsyl``` and ```dtgsyl``` for a dynamic system analyzing package.

### Describe the solution you'd like.

A low-level wrapper in the sub-package [scipy.linalg.lapack](https://docs.scipy.org/doc/scipy/reference/linalg.lapack.html#module-scipy.linalg.lapack) would be sufficient.

### Describe alternatives you've considered.

Currently I use the [python-julia interface](https://pypi.org/project/julia/) and the [sylvsys function](https://docs.juliahub.com/MatrixEquations/1uOBF/1.2.0/sylvester.html) in the MatrixEquations.jl package.
However, a pure Python solution would be advantageous.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by mainak33**
Reference issue

Closes #18404 

What does this implement/fix?

This adds a low-level wrapper for Lapack functions <s/d>tgsyl

Additional information:
- Not final code. Help and review welcome
- In some tests I tried to calculate lhs and rhs of the sylvester equations after solution, but they don't seem to meet the desired accuracy of machine precision*100 that is used in other tests. So, I still have to debug. 
- For now I have these questions: 
>1) I'm not too sure what the input matrix forms should be. Lapack doc does say (a,b) and (d,e) should be quasi upper triangular and upper triangular respectively. Does quasi upper triangular mean block upper triangular? 
>2) How do we check the matrix forms like (quasi) triangular inside the template code? 
>3) I am still not sure how f2py is actually building the *.pyf.src template signature files. Does f2py itself have the mechanism to replace the templates or is there some extra build step somewhere to generate the actual *.pyf from the templates? 
>4) If you are aware of other sources for tests let me know (NAG is mentioned in code. But not sure where to find it.)

### CODE REVIEW

File: scipy/linalg/flapack_other.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_other.pyf.src:407]: Probably `0` as the default mode is better.

File: scipy/linalg/flapack_other.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_other.pyf.src:433]: iwork is also a scratch register you can hide it.

File: scipy/linalg/flapack_other.pyf.src
Comment (ilayn) [inline scipy/linalg/flapack_other.pyf.src:438]: I've read the code a bit more and indeed we don't need this query. My sincere apologies, I thought it was doing something special but it is not

```fortran
*
      IF( info.EQ.0 ) THEN
         IF( notran ) THEN
            IF( ijob.EQ.1 .OR. ijob.EQ.2 ) THEN
               lwmin = max( 1, 2*m*n )
            ELSE
               lwmin = 1
            END IF
         ELSE
            lwmin = 1
         END IF
         work( 1 ) = lwmin
```

In fact for ijob 0 or 3 it is only using 1-long `lwork`, so we can use this and keep lwork as optional indeed.

**Review by ilayn — CHANGES_REQUESTED** [review]
Thanks @mainak33 please see my comments sorry again for not looking at it earlier about `lwork`.

File: scipy/linalg/flapack_other.pyf.src
Comment (mainak33) [reply scipy/linalg/flapack_other.pyf.src:438]: Thanks for looking into it. I was stuck on getting lwork =1 in the tests and thought there was something wrong with my code. I will remove the _lwork routine and make lwork optional in *tgsyl.

File: scipy/linalg/tests/test_lapack.py
Comment (mainak33) [inline scipy/linalg/tests/test_lapack.py:3352]: @ilayn @Kai-Striega I can't see why the precision is so poor. I can of course, increase atol and rtol further or compare results upto certain decimals, but I'm a bit circumspect in doing so. Let me know if you see anything obvious in the test code. The only thing which maybe causing a problem is that the element(0,0) of a and b are larger compared to others since I used schur to get them into the quasi triangular form. An example A matrix looks like this: 
image

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [inline scipy/linalg/tests/test_lapack.py:3344]: We try to avoid using `assert_almost_equal` and some similar functions in new code. See [Code and Documentation Style Guide - The Missing Bits](https://docs.scipy.org/doc/scipy/dev/missing-bits.html#test-functions-from-numpy-testing) for some more info.

Comment (mainak33) [reply scipy/linalg/tests/test_lapack.py]: Addressed it in my next checkin. Also increased tolerance for single precision. Let's see if tests pass.

**Comment by ilayn** [comment]
Looks almost done to me. However the test need a full PEP8 pass then this is getting ready for merge.

**Comment by mainak33** [comment]
> Looks almost done to me. However the test need a full PEP8 pass then this is getting ready for merge.
Fixed PEP8 styling issues with new tests in latest checkin

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [inline scipy/linalg/tests/test_lapack.py:3380]: Here and below the `if` statements don't require parentheses around them

**Review by Kai-Striega — COMMENTED** [review]
Looks much better after the PEP8 changes. Thank you

File: scipy/linalg/tests/test_lapack.py
Comment (mainak33) [reply scipy/linalg/tests/test_lapack.py]: Fixed in latest commit. Thanks!

Comment (mainak33) [reply scipy/linalg/tests/test_lapack.py]: The failing tests seem unrelated to my changes. I just rebased after removing parantheses and pushed.

**Review by Kai-Striega — COMMENTED** [review]
I'm happy with your work, and I believe the failing tests are unrelated to your changes.  Well done @mainak33! Let's wait for a final pass by @ilayn and then it should be fine

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [inline scipy/linalg/tests/test_lapack.py:3363]: I'm still not too sure why the tolerance needs to be this low. Maybe there should be some investigation why this is happening?

File: scipy/linalg/lapack.py
Comment (Kai-Striega) [inline scipy/linalg/lapack.py:662]: Random whitespace should be taken out

Comment (mainak33) [reply scipy/linalg/lapack.py]: Removed in latest commit/rebase

File: scipy/linalg/tests/test_lapack.py
Comment (mainak33) [reply scipy/linalg/tests/test_lapack.py:3363]: Just to be on the same page, I think you mean the tolerance is too high (too loose). I agree. I think the only thing I can think of checking here would be if we are loosing precision between fortran lapack output and python. The fortran wrapper is being constructed by f2py and I am not sure how to debug it directly. Will write statements in the template signature file work? Or is there some more sophisticated debug option I am not aware of? @ilayn Please also weigh in if you have thoughts.

Comment (mainak33) [reply scipy/linalg/tests/test_lapack.py:3363]: Alternatively, I thought of using the same tests Lapack has. I see they test stgsyl here: https://github.com/Reference-LAPACK/lapack/blob/f86d08bae6c46685e0f759850e82dc0a8fa00a21/TESTING/EIG/serrgg.f#LL1131C4-L1131C4
But it looks like they are using 1x1 matrices or testing error with dimensions maybe.

Comment (Kai-Striega) [reply scipy/linalg/tests/test_lapack.py:3363]: It might be worth doing. But, if it's needed, can probably be done in another PR. TBH I'm not a LAPACK expert so I'm going to defer to what @ilayn thinks.

Comment (ilayn) [reply scipy/linalg/tests/test_lapack.py:3363]: I have one proper idea and one hunch, because we are getting data from random arrays, their spectrum is probably too close to have a decent solution hence could you please add `5*np.eye(m, dtype=dtype)` to `a` and subtract the same from `d`? 

Also (hunch) could you please set `scale` to `1` in the wrapper? I feel like it gets referenced without being set. If locally it passes with proper tolerance values (around `1e-10` should be ok for double precision) then it is OK.

Comment (mainak33) [reply scipy/linalg/tests/test_lapack.py:3363]: @ilayn 
> could you please add 5*np.eye(m, dtype=dtype) to a and subtract the same from d?

Thanks for your insight! I tried this and on my system it improved things a lot. I could go down to any atol for double precision (1e-10 works). For single precision atol =2e-3 works but 1e-3 fails. From what I have seen other systems used  in the CI here need somewhat looser tolerances than my system. I'll try with 1e-10 and 5e-3 for my next checkin and see how it goes.

> Also (hunch) could you please set scale to 1 in the wrapper?

Currently, the intent for the scale argument in the wrapper is out. Do you want me to make the intent in,out or inout? Will wait for your response before making changes

**Comment by ilayn** [comment]
If you separate the spectrum a bit further then float32 should also get ~1e-4 level precision. So something like this should make it fly. Nevermind the `scale` comment above. 

```python
    atol = 1e-4 if dtype == np.float32 else 1e-10

    m, n = 10, 15

    a, _ = schur(generate_random_dtype_array([m, m], dtype=dtype))
    d = np.triu(generate_random_dtype_array([m, m], dtype=dtype))
    b, _ = schur(generate_random_dtype_array([n, n], dtype=dtype))
    e = np.triu(generate_random_dtype_array([n, n], dtype=dtype))
    c = generate_random_dtype_array([m, n], dtype=dtype)
    f = generate_random_dtype_array([m, n], dtype=dtype)

    # Adjust the spectrum of the matrices to get a more precise solution
    adjustment = np.eye(m, dtype=dtype)
    a += adjustment
    d -= 5 * adjustment
    b += adjustment
    e -= 5 * adjustment
```

**Comment by mainak33** [comment]
> > If you separate the spectrum a bit further then float32 should also get ~1e-4 level precision. So something like this should make it fly...
> 
> I had to make a few minor changes to your suggested code due to matrix sizes but it worked on my local machine. I pushed it. Let's see if it passes all CI tests.

Tests are failing on some platforms for atol=1e-4 for single precision. Trying with atol =1e-3

**Comment by ilayn** [comment]
Well these are tricky problems anyways so if it works with 1e-3 it is what it is, no need to dwell on it. If things pass, let's get it in instead of wasting your time.

**Comment by ilayn** [comment]
Ah OK, the transposed problem is a different one to separate, hence the artificial push is making the `'T'` option solution worse. That's why we get that silly result in float32 probably. Almost there but I'll check this tomorrow.

**Comment by mainak33** [comment]
> Ah OK, the transposed problem is a different one to separate, hence the artificial push is making the `'T'` option solution worse. That's why we get that silly result in float32 probably. Almost there but I'll check this tomorrow.

Thanks for looking into it. I'm afraid I don't really know the specifics of this algorithm, so won't be much help figuring out the tolerances beyond trial and error. I'll  leave it upto you to fine tune the test and tolerances. Let me know if I can help otherwise or if I need to take any action once we are ready to merge the PR.

**Comment by mainak33** [comment]
The one failing test on macOS tests (meson) / Meson build (3.11) (https://github.com/scipy/scipy/actions/runs/5162854702/jobs/9300833616?pr=18571) looks like its related to  #18601 ?

**Comment by ilayn** [comment]
Ah dammit. I'm stupid. That's why indeed. Anyways let's wait for that one. This one is indeed ready with smaller tolerances

**Comment by mainak33** [comment]
> Ah dammit. I'm stupid. That's why indeed. Anyways let's wait for that one. This one is indeed ready with smaller tolerances

No worries. I agree. This one looks ready pending the build fix for macOS. Thanks for all the help!

**Comment by mainak33** [comment]
@ilayn @Kai-Striega I just rebased and pushed and it is now passing all tests with tighter tolerances that Ilayn worked on (Looks like the BLAS package updates for #18601 must have worked to fix the OSX tests for this as well). I think this is now ready for merge.

**Review by Kai-Striega — APPROVED** [review]
I'm happy with this. Waiting on @ilayn for a final review and merge.

**Comment by ilayn** [comment]
Looks good, thank you very much @mainak33 for your patience and @Kai-Striega for the reviews. Let's click it

**Pull request merged**
